activestorage/test/models/preview_test.rb

```diff
@@ -72,6 +74,56 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
     end
   end
 
+  test "#processed does not generate a variant when provided an empty transformation" do
+    blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+    preview = blob.preview({})
+
+    preview.processed
+
+    assert_empty preview.image.variant_records
+  end
+
+  PREVIEW_METHODS.each do |method_under_test|
+    test "##{method_under_test} does not generate a variant when provided an empty transformation" do
+      freeze_time do
+        blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+        empty_transformation = {}
+        preview = blob.preview(empty_transformation)
+
+        assert_no_changes -> { !!preview.image.variant(empty_transformation)&.send(:processed?) } do
+          preview.processed.send(method_under_test)
+        end
+        assert_equal blob.preview_image.send(method_under_test), preview.processed.send(method_under_test)
+      end
+    end
+  end
+
+  PREVIEW_METHODS.each do |method_under_test|
+    test "##{method_under_test} returns the variant's #{method_under_test} when a transformation is present" do
+      freeze_time do
+        blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+        preview = blob.preview(resize_to_limit: [640, 280])
+        preview.processed
+        expected_result = preview.image.variant_records[0].image.send(method_under_test)
+
+        actual_result = preview.send(method_under_test)
+
+        assert_equal expected_result, actual_result
+      end
+    end
+  end
+
+  PREVIEW_METHODS.each do |method_under_test|
+    test "#{method_under_test} raises `UnprocessedError` when preview is not processed" do
+      blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+      preview = blob.preview(resize_to_limit: [640, 280])
+
+      assert_raises ActiveStorage::Preview::UnprocessedError do
+        preview.send(method_under_test)
+      end
+    end
+  end
```

Comment:
Declaring tests in a loop like this can be a good strategy for exhaustive test coverage.  However, in this case, I think it adds unnecessary complexity.  It is also a leaky abstraction.  `PREVIEW_METHODS` does not accommodate methods that require arguments.  So if we were to add such methods in the future, we would need to rework this entire approach.

I believe we can cover all the behavior we're interested in with the following two tests:

```ruby
test "image-related methods raise UnprocessedError when preview is not processed" do
  blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
  preview = blob.preview(resize_to_limit: [640, 280])

  assert_raises(ActiveStorage::Preview::UnprocessedError) { preview.url }
  assert_raises(ActiveStorage::Preview::UnprocessedError) { preview.key }
  assert_raises(ActiveStorage::Preview::UnprocessedError) { preview.download }
end

test "previewing with empty transformations does not generate a variant" do
  blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
  preview = blob.preview({})

  preview.processed

  freeze_time { assert_equal blob.preview_image.url, preview.url }
  assert_equal blob.preview_image.key, preview.key
  assert_equal blob.preview_image.download, preview.download

  assert_empty preview.image.variant_records
end
```

Let's use those instead.

Reply:
Done. thanks